### PR TITLE
Issue #13: Dynamic Signers

### PR DESCRIPTION
**Pull request by valeriomazzeo**
This pull request aims to implement what's described in #13.

It depends from:

- https://github.com/vapor/jwt/pull/61
- https://github.com/vapor/jwt/pull/62

Once those are merged the JWT package can be updated and the following files can be removed from this branch:

- `RSAKey+JWK.swift`
- `KeyIDHeader.swift`

**Note:** It will require a minor version update of the `JWTProvider` package.

## Configuration

`JWTProvider` can be configured in **3** different ways:

- Custom signers defined in `jwt.json`
  - Supports (private/public): `hmac`, `rsa`, `esdca`.
- Legacy custom signer defined in `jwt.json`. 
  - Supports (private/public): `hmac`, `rsa`, `esdca`.
- Remote JSON Web Key Set (`jwks.json`) URL
  - Supports (private/public): `rsa`.

If your Vapor app is acting as an Authentication Provider, you may want to use either the `Legacy custom signer` setup, or the `Custom signers` setup, which is great if you want to perform certificates rotation.

The only difference is that with `Custom signers` the `kid` value in the `JWT` header is not ignored, and it must match an associated signer in order to verify the signature.

If your Vapor app is a Resource Provider that delegates Authentication to a 3rd party (auth0, stormpath, etc), you may want to use the `Remote JSON Web Key Set` setup. In this configuration the JWT token is generated by a 3rd party that provides the public key in JSON Web Key Set format.
The Vapor app is only in charge to verify the `JWT` signature using the key set provided by the 3rd party.

#### Remote JSON Web Key Set

**jwt.json**
```json
{
  "jwks-url": "http://my-domain.com/well-known/jwks.json"
}
```

#### Custom Signers

This allows to specify an array of signers and is particularly useful for rotating certificates.
Custom signers are not backward compatible and must specify an additional `kid` in the configuration.

- type: `unsigned`, `hmac`, `rsa`, `esdca`
- kid: an unique identifier
- algorithm:
  - type[`hmac`]: `hs256`, `hs384`, `hs512`
  - type[`rsa`]: `rs256`, `rs384`, `rs512`
  - type[`esdca`]: `es256`, `es384`, `es512`

**jwt.json**
```json
{
  "signers": [
    {
      "type": "rsa",
      "kid": "1234",
      "algorithm": "rs256",
      "key": "yourkeyhere"
    }
  ]
}
```

#### Legacy Custom Signer

This is backwards compatible with the previous implementation.

- type: `unsigned`, `hmac`, `rsa`, `esdca`
- algorithm:
  - type[`hmac`]: `hs256`, `hs384`, `hs512`
  - type[`rsa`]: `rs256`, `rs384`, `rs512`
  - type[`esdca`]: `es256`, `es384`, `es512`

**jwt.json**
```json
{
  "signer": {
    "type": "rsa",
    "algorithm": "rs256",
    "key": "yourkeyhere"
  }
}
```

## TODO:

 - [x] `n` and `e` in `JSONWebKeySignerFactory`
- [ ] ~Support `x5c`~
- [ ] Rate Limiting when retrieving a new `jwks.json`
- [x] Unit test

### CODE REVIEW

File: Sources/JWTProvider/Droplet+JWT.swift [hunk 1]
```diff
@@ -1,16 +1,27 @@
...
 import JWT
 
-private let jwtSignerKey = "jwt-provider:signer"
-
 extension Droplet {
+    @available(*, deprecated, message: "Use signers instead.")
     public internal(set) var signer: Signer? {
-        get { return storage[jwtSignerKey] as? Signer }
-        set { storage[jwtSignerKey] = newValue }
+        get { return self.signers?["_legacy"] }
```
Comment (siemensikkema) [inline Sources/JWTProvider/Droplet+JWT.swift]: Please use a constant (e.g. `jwtLegacySignerKey`)

File: Sources/JWTProvider/Droplet+JWT.swift [hunk 2]
```diff
@@ -19,3 +30,41 @@ extension Droplet {
...
+    public func assertSigners() throws -> SignerMap {
+        guard let signers = self.signers else {
+            throw JWTProviderError.noJWTSigner
+        }
+
+        return signers
+    }
+}
+
+private let jwtJWKSURL = "jwt-provider:jwks-url"
```
Comment (siemensikkema) [inline Sources/JWTProvider/Droplet+JWT.swift]: `jwtJWKSURLKey`

File: Sources/JWTProvider/PayloadAuthenticationMiddleware.swift [hunk 3]
```diff
@@ -8,19 +8,46 @@ import AuthProvider
...
 public final class PayloadAuthenticationMiddleware<U: PayloadAuthenticatable>: Middleware {
-    let signer: Signer
+    internal private(set) var signers: SignerMap
     let claims: [Claim]
+    let jwksURL: String?
+    let clientFactory: ClientFactoryProtocol?
 
     /// Create a LoginMiddleware specifying
-    /// the JWT signer and type of payload
+    /// the JWT signers and type of payload
```
Comment (siemensikkema) [inline Sources/JWTProvider/PayloadAuthenticationMiddleware.swift]: Header doc does not match init signature (signers vs. signer)

File: Sources/JWTProvider/PayloadAuthenticationMiddleware.swift [hunk 4]
```diff
@@ -8,19 +8,46 @@ import AuthProvider
...
     /// that will be stored in the JWT
     public init(
         _ signer: Signer,
         _ claims: [Claim] = [],
         _ userType: U.Type = U.self
+        ) {
```
Comment (siemensikkema) [inline Sources/JWTProvider/PayloadAuthenticationMiddleware.swift]: indentation

File: Sources/JWTProvider/PayloadAuthenticationMiddleware.swift [hunk 5]
```diff
@@ -8,19 +8,46 @@ import AuthProvider
...
+        self.jwksURL = nil
+        self.clientFactory = nil
+    }
+
+    public init(
+        _ jwksURL: String,
+        _ claims: [Claim] = [],
+        _ userType: U.Type = U.self,
+        clientFactory: ClientFactoryProtocol = EngineClientFactory()
+        ) {
```
Comment (siemensikkema) [inline Sources/JWTProvider/PayloadAuthenticationMiddleware.swift]: indentation

File: Sources/JWTProvider/SignerMap.swift [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+        var map = [String: Signer]()
+
+        for key in keys {
+
+            guard let kid: String = try key.get("kid") else {
+                continue
+            }
+
+            guard let signer = try? JSONWebKeySignerFactory(jwk: key).makeSigner() else {
```
Comment (siemensikkema) [inline Sources/JWTProvider/SignerMap.swift]: can be combined with previous guard statement

File: Sources/JWTProvider/JSONWebKeySignerFactory.swift [hunk 7]
```diff
@@ -0,0 +1,67 @@
+//
```
Comment (siemensikkema) [inline Sources/JWTProvider/JSONWebKeySignerFactory.swift]: These file headers are not normally part of vapor source code

File: Sources/JWTProvider/PayloadAuthenticationMiddleware.swift (hunk 4 above)
Comment (valeriomazzeo) [reply Sources/JWTProvider/PayloadAuthenticationMiddleware.swift]: if you press `CTRL+I` it returns to that...

File: Sources/JWTProvider/PayloadAuthenticationMiddleware.swift (hunk 5 above)
Comment (valeriomazzeo) [reply Sources/JWTProvider/PayloadAuthenticationMiddleware.swift]: same as above

File: Sources/JWTProvider/JSONWebKeySignerFactory.swift (hunk 7 above)
Comment (valeriomazzeo) [reply Sources/JWTProvider/JSONWebKeySignerFactory.swift]: Ok, I can remove the headers

File: Sources/JWTProvider/PayloadAuthenticationMiddleware.swift (hunk 4 above)
Comment (siemensikkema) [reply Sources/JWTProvider/PayloadAuthenticationMiddleware.swift]: I know, it's annoying! I think the consensus is that we sometimes ignore Xcode when it comes to formatting.

File: Sources/JWTProvider/SignerMap.swift (hunk 6 above)
Comment (valeriomazzeo) [reply Sources/JWTProvider/SignerMap.swift]: I personally find it less readable, but as you wish...

File: Sources/JWTProvider/PayloadAuthenticationMiddleware.swift (hunk 4 above)
Comment (valeriomazzeo) [reply Sources/JWTProvider/PayloadAuthenticationMiddleware.swift]: Can you please show me how you want it indented?

Comment (siemensikkema) [reply Sources/JWTProvider/PayloadAuthenticationMiddleware.swift]: `) {` should be aligned with `public`

Comment (valeriomazzeo) [reply Sources/JWTProvider/PayloadAuthenticationMiddleware.swift]: it's done already

**Comment by siemensikkema** [comment]
@LoganWright The breaking change label can be removed now that @valeriomazzeo has addressed it with the deprecation warnings right? Or does deprecation still count as a breaking change?

**Comment by valeriomazzeo** [comment]
Hi guys, @siemensikkema @LoganWright @tanner0101 

Is there any chance we can get this moving?

File: Sources/JWTProvider/Droplet+JWT.swift [hunk 8]
```diff
@@ -1,16 +1,29 @@
 import Vapor
 import JWT
 
-private let jwtSignerKey = "jwt-provider:signer"
+let jwtLegacySignerKey = "jwtLegacySignerKey"
```
Comment (tanner0101) [inline Sources/JWTProvider/Droplet+JWT.swift]: `jwt-providers:legacy-signer` would be better. the namespace is helpful.

File: Sources/JWTProvider/JWTProviderError.swift [hunk 9]
```diff
@@ -4,12 +4,13 @@ import HTTP
 public enum JWTProviderError: Error {
     case noVerifiedJWT
     case noJWTSigner
+    case noJWKSURL
```
Comment (tanner0101) [inline Sources/JWTProvider/JWTProviderError.swift]: Adding cases to an enum is a breaking change. We'll have to add a new error type or use one of the existing ones.

File: Sources/JWTProvider/Provider.swift [hunk 10]
```diff
@@ -4,108 +4,45 @@ import JWT
 /// Adds required JWT objects to your application
 /// like token Signers
 public final class Provider: Vapor.Provider {
+
     public static let repositoryName = "jwt-provider"
     
-    public let signer: Signer
```
Comment (tanner0101) [inline Sources/JWTProvider/Provider.swift]: Looks like this method and the init under it were removed w/o a deprecation, unless I missed it.

File: Sources/JWTProvider/SignerMap.swift [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+            throw SignerMapError.missingKey("keys")
+        }
+
+        var map = [String: Signer]()
+
+        for key in keys {
+
+            guard let kid: String = try key.get("kid"),
+                let signer = try? JSONWebKeySignerFactory(jwk: key).makeSigner()
+                else {
```
Comment (tanner0101) [inline Sources/JWTProvider/SignerMap.swift]: formatting

**Review by tanner0101 — CHANGES_REQUESTED** [review]
This looks great. Just a couple of minor issues that need to be addressed before we can merge.

Additionally, we will need some updated documentation about how to use these new features and ideally a bit of information about how to upgrade your deprecated code ( @siemensikkema deprecated code should be fine as long as its a minor version bump).

@valeriomazzeo do you think you could include as a comment here some markdown documenting the new API and some additional notes about how to migrate existing code? I can take the markdown from here and put it up on docs.vapor.codes.

**Comment by valeriomazzeo** [comment]
@tanner0101 detailed documentation is in the PR, I have already suggested to @LoganWright to publish it in docs.vapor.codes once this is merged.

Do you want to add it to the source code as well? I am not sure where, it would probably be faster if you sort it out the way you like it.

File: Sources/JWTProvider/SignerMap.swift (hunk 11 above)
Comment (valeriomazzeo) [reply Sources/JWTProvider/SignerMap.swift]: do you prefer the `else` to be aligned with guard? Xcode keeps moving it ...

**Comment by dlbuckley** [comment]
@tanner0101 @valeriomazzeo whats the latest on this PR?

**Review by tanner0101 — APPROVED** [review]

**Comment by tanner0101** [comment]
Thanks for the hard work done here @valeriomazzeo :). Merging this now and I'm going to test it with some large projects I have, update the docs, then release it!

**Pull request merged**